### PR TITLE
🐛 Restore v1alpha3/v1alpha4 conversion to fix SSA issue & add e2e test coverage

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

This PR:
* Fixes the issue with old apiVersions in managedFields + SSA by restoring conversion of v1alpha3 & v1alpha4
* Adds corresponding e2e test coverage to verify the fix works and avoid regressions

More details on the issue soon.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/10051

### CODE REVIEW

**Review by fabriziopandini — COMMENTED** [review]
Awesome! 
This is a big PR and a lot of work in such a short time and I would like to praise a few things make this PR to stand out
- Restoring APIs in internal packages really helps in keeping the public API surfaces as minimal and possible
- Adding E2e test validating a double clusterctl upgrade allows us to make this change with confidence (and we will benefit from this as well down the line when working to a more sophisticated solution)
- The trick to force controller manager restart is a real piece of art

just a few nits, nothing really blocking

File: Makefile [hunk 1]
```diff
@@ -441,9 +441,22 @@ generate-go-conversions-core: ## Run all generate-go-conversions-core-* targets
...
+	$(CONVERSION_GEN) \
+		--input-dirs=./api/v1alpha4 \
+		--build-tag=ignore_autogenerated_core \
+		--output-file-base=zz_generated.conversion $(CONVERSION_GEN_OUTPUT_BASE) \
+		--go-header-file=./hack/boilerplate/boilerplate.generatego.txt
 
 .PHONY: generate-go-conversions-core-exp
 generate-go-conversions-core-exp: $(CONVERSION_GEN) ## Generate conversions go code for core exp
+	$(MAKE) clean-generated-conversions SRC_DIRS="./$(EXP_DIR)/api/v1alpha4,./$(EXP_DIR)/addons/api/v1alpha4"
+	$(CONVERSION_GEN) \
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply Makefile:458]: Nice catch. I'll add it

File: test/framework/clusterctl/client.go [hunk 2]
```diff
@@ -218,6 +189,79 @@ func Upgrade(ctx context.Context, input UpgradeInput) {
...
+	if err != nil {
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			stdErr = string(exitErr.Stderr)
+		}
+	}
+	Expect(err).ToNot(HaveOccurred(), "failed to run clusterctl upgrade apply:\nstdout:\n%s\nstderr:\n%s", string(out), stdErr)
+}
+
+func calculateClusterCtlUpgradeArgs(input UpgradeInput) []string {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/framework/clusterctl/client.go:234]: Nope. There are two funcs calculateClusterCtlUpgradeArgs & calculateClusterCtlInitArgs

File: test/e2e/clusterctl_upgrade_test.go [hunk 3]
```diff
@@ -34,6 +36,60 @@ var (
 	providerDockerPrefix  = "docker:v%s"
 )
 
+var _ = Describe("When testing clusterctl upgrades (v0.4=>v1.6=>current)", func() {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade_test.go:107]: Will add one

**Comment by fabriziopandini** [comment]
/lgtm
/approve
Feel free to unwip/unhold once the test are completed

**Comment by sbueringer** [comment]
> I haven't done a full review of this -but I'm trying to understand the change and the root cause.
> 
> Do we need to keep apiVersions around pretty much indefinitely - but with `served: false` in order to keep the conversion webhooks working?
> 
> At least until we can be _sure_ that all apiVersions in the managed fields have been updated?

Yup. That's exactly it on a high-level. Fabrizio will share more details in the office hours (please check the agenda the information is already there, we'll update the issue later accordingly as well).

This PR and its cherry-pick are only meant as a temporary solution until we figure out the right way to handle this problem. The idea was to only restore the minimal amount necessary to get conversion working again (which is the reason why we added v1alpha3/v1alpha4 under internal).

**Comment by sbueringer** [comment]
P.S. 2 Sorry for the short time frame to review. Trying to get this fixed up before I'm going on PTO on Friday and in time so we have some time in CI before the 1.6 patch release.

**Pull request merged**